### DISCUSSION TITLE
Update embedded SemVersion class or use Semver package again

### DISCUSSION THREAD

**Discussion by ronaldbarendse** [umbraco/Umbraco-CMS#14539] [Features and ideas]
Umbraco 8 still had a package reference on `Semver`:
https://github.com/umbraco/Umbraco-CMS/blob/159f92d6f846846a356a461aa7114e129d90e5ba/src/Umbraco.Core/Umbraco.Core.csproj#L95

But in Umbraco 9, this reference has been removed and the `SemVersion` class was copied over, see commit: https://github.com/umbraco/Umbraco-CMS/commit/152ba318bf030264a40aad1a5ab74a99e107c594. I believe the motivation was to only have .NET (System) and Microsoft dependencies in the core Umbraco projects.

We updated this class several times to add nullability support and in a recent PR I did, I've added an additional attribute to improve the nullable static analysis:
https://github.com/umbraco/Umbraco-CMS/blob/84a0cd8c091a4c8de17fcd93176eb2f373866497/src/Umbraco.Core/Semver/Semver.cs#L199

The Semver package is currently on version 2.3.0, so there are potentially some bugfixes we haven't included yet. The reverse also seems to be the case: we've added nullable support, but that's something not yet available in the latest Semver release (although the source seems to be updated and this will probably be released in the next major).

So what should we do?
1. Nothing, because it works fine now 🤡;
2. Keep the embedded `SemVersion` class (potentially lagging behind again in the future):
  a. Contribute our changes back to the original repository and create a copy again (making it easier to stay up-to-date later);
  b. Update the class from the original repository and re-apply any changes we did (in case we can't contribute them back).
3. Remove the embedded class and take a dependency on the package again (in v13, since it's a breaking change and by that time a new major version with nullable support might be released).

**Comment by bergmania** [comment]
IMO we are too tightly bound to that class today. 

We only use a subset of the features and still have quite some extension methods.

I guess it would be fairly simple to just have our own class instead

**Reply by bergmania** [reply]
Alternatively we should just use a simple POCO for the 99% of the cases and then have a `ISemVersionFactory` that is implemented in Infrastructure by using the `SemVer` package. Thereby we could rely on the functionality to do the parsing of versions.

**Reply by bergmania** [reply]
[This](https://github.com/umbraco/Umbraco-CMS/pull/14755) could work, but is my the nature of it breaking.
Especially ModelsBuilder have breaking changes, because it used alot of static stuff.
